src/solvers/refinement/string_constraint_generator_float.cpp

```diff
@@ -0,0 +1,445 @@
+///   sufficient for the exponent even in octuple precision.
+exprt get_exponent(
+  const exprt &src, const ieee_float_spect &spec)
+{
+  exprt exp_bits=extractbits_exprt(
+    src, spec.f+spec.e-1, spec.f, unsignedbv_typet(spec.e));
+
+  // Exponent is in biased from (numbers form -128 to 127 are encoded with
+  // integer from 0 to 255) we have to remove the bias.
+  return minus_exprt(
+    typecast_exprt(exp_bits, signedbv_typet(32)),
+    from_integer(spec.bias(), signedbv_typet(32)));
+}
+
+/// Gets the magnitude without hidden bit
+/// \param src: a floating point expression
+/// \param spec: specification for floating points
+/// \return An unsigned value representing the magnitude.
+exprt get_magnitude(const exprt &src, const ieee_float_spect &spec)
+{
+  return extractbits_exprt(src, spec.f-1, 0, unsignedbv_typet(spec.f));
+}
+
+/// Gets the significand as a java integer, looking for the hidden bit
+/// \param src: a floating point expression
+/// \param spec: specification for floating points
+/// \param type: type of the output, should be unsigned with width greater than
+///   the width of the significand in the given spec
+/// \return An integer expression of the given type representing the
+///   significand.
+exprt get_significand(
+  const exprt &src, const ieee_float_spect &spec, const typet &type)
+{
+  assert(type.id()==ID_unsignedbv);
+  assert(to_unsignedbv_type(type).get_width()>spec.f);
+  typecast_exprt magnitude(get_magnitude(src, spec), type);
+  exprt exponent=get_exponent(src, spec);
+  equal_exprt all_zeros(exponent, from_integer(0, exponent.type()));
+  exprt hidden_bit=from_integer((1 << spec.f), type);
+  plus_exprt with_hidden_bit(magnitude, hidden_bit);
+  return if_exprt(all_zeros, magnitude, with_hidden_bit);
+}
+
+/// \param f: a floating point value in double precision
+/// \return an expression representing this floating point
+exprt constant_float(const double f, const ieee_float_spect &float_spec)
+{
+  ieee_floatt fl(float_spec);
+  if(float_spec==ieee_float_spect::single_precision())
+  {
+    fl.from_float(f);
+  }
+  else
+  {
+    assert(float_spec==ieee_float_spect::double_precision());
+    fl.from_double(f);
+  }
+  return fl.to_expr();
+}
+
+/// Round a float expression to an integer closer to 0
+/// \param f: expression representing a float
+/// \return expression representing a 32 bit integer
+exprt round_expr_to_zero(const exprt &f)
+{
+  exprt rounding=from_integer(ieee_floatt::ROUND_TO_ZERO, unsignedbv_typet(32));
+  return floatbv_typecast_exprt(f, rounding, signedbv_typet(32));
+}
+
+/// Multiplication of expressions representing floating points.
+/// Note that the rounding mode is set to ROUND_TO_EVEN.
+/// \param f: a floating point expression
+/// \param g: a floating point expression
+/// \param rounding: rounding mode
+/// \return An expression representing floating point multiplication.
+exprt floatbv_mult(const exprt &f, const exprt &g)
```

Comment:
For this particular function it is used in every calls of  add_axioms_from_float_scientific_notation
so regression test for that functionality should also cover that function.